README.md
# rsparse <img src='man/figures/logo.png' align="right" height="128" />
<!-- badges: start -->
[![R build status](https://github.com/rexyai/rsparse/workflows/R-CMD-check/badge.svg)](https://github.com/rexyai/rsparse/actions)
[![codecov](https://codecov.io/gh/rexyai/rsparse/branch/master/graph/badge.svg)](https://codecov.io/gh/rexyai/rsparse/branch/master)
[![License](https://eddelbuettel.github.io/badges/GPL2+.svg)](http://www.gnu.org/licenses/gpl-2.0.html)
[![Project Status](https://img.shields.io/badge/lifecycle-maturing-blue.svg)](https://lifecycle.r-lib.org/articles/stages.html#maturing)
<a href="https://rexy.ai"><img src="https://s3-eu-west-1.amazonaws.com/rexy.ai/images/favicon.ico" height="32" width="32"></a>
<!-- badges: end -->

`rsparse` is an R package for statistical learning primarily on **sparse matrices** -  **matrix factorizations, factorization machines, out-of-core regression**. Many of the implemented algorithms are particularly useful for **recommender systems** and **NLP**. 

We've paid some attention to the implementation details - we try to avoid data copies, utilize multiple threads via OpenMP and use SIMD where appropriate. Package **allows to work on datasets with millions of rows and millions of columns**.

# Features

### Classification/Regression

1. [Follow the proximally-regularized leader](http://proceedings.mlr.press/v15/mcmahan11b/mcmahan11b.pdf) which allows to solve **very large linear/logistic regression** problems with elastic-net penalty. Solver uses stochastic gradient descent with adaptive learning rates (so can be used for online learning - not necessary to load all data to RAM). See [Ad Click Prediction: a View from the Trenches](https://www.eecs.tufts.edu/~dsculley/papers/ad-click-prediction.pdf) for more examples.
    - Only logistic regerssion implemented at the moment
    - Native format for matrices is CSR - `Matrix::RsparseMatrix`. However common R `Matrix::CsparseMatrix` (`dgCMatrix`) will be converted automatically.
1. [Factorization Machines](https://www.csie.ntu.edu.tw/~b97053/paper/Rendle2010FM.pdf) supervised learning algorithm which learns second order polynomial interactions in a factorized way. We provide highly optimized SIMD accelerated implementation.  

### Matrix Factorizations

1. Vanilla **Maximum Margin Matrix Factorization** - classic approch for "rating" prediction. See `WRMF` class and constructor option `feedback = "explicit"`. Original paper which indroduced MMMF could be found [here](https://ttic.uchicago.edu/~nati/Publications/MMMFnips04.pdf).
    * <img src="https://raw.githubusercontent.com/rexyai/rsparse/master/docs/img/MMMF.png" width="400">
1. **Weighted Regularized Matrix Factorization (WRMF)** from [Collaborative Filtering for Implicit Feedback Datasets](http://yifanhu.net/PUB/cf.pdf). See `WRMF` class and constructor option `feedback = "implicit"`. 
We provide 2 solvers:
    1. Exact based on Cholesky Factorization
    1. Approximated based on fixed number of steps of **Conjugate Gradient**.
See details in [Applications of the Conjugate Gradient Method for Implicit Feedback Collaborative Filtering](http://www.sze.hu/~gtakacs/download/recsys_2011_draft.pdf) and [Faster Implicit Matrix Factorization](http://www.benfrederickson.com/fast-implicit-matrix-factorization/).
    * <img src="https://raw.githubusercontent.com/rexyai/rsparse/master/docs/img/WRMF.png" width="400">
1. **Linear-Flow** from [Practical Linear Models for Large-Scale One-Class Collaborative Filtering](http://www.bkveton.com/docs/ijcai2016.pdf). Algorithm looks for factorized low-rank item-item similarity matrix (in some sense it is similar to [SLIM](http://glaros.dtc.umn.edu/gkhome/node/774))
    * <img src="https://raw.githubusercontent.com/rexyai/rsparse/master/docs/img/LinearFlow.png" width="300">
1. Fast **Truncated SVD** and **Truncated Soft-SVD** via Alternating Least Squares as described in [Matrix Completion and Low-Rank SVD via Fast Alternating Least Squares](https://arxiv.org/pdf/1410.2596.pdf). Works for both sparse and dense matrices. Works on [float](https://github.com/wrathematics/float) matrices as well! For certain problems may be even faster than [irlba](https://github.com/bwlewis/irlba) package.
    * <img src="https://raw.githubusercontent.com/rexyai/rsparse/master/docs/img/soft-svd.png" width="600">
1. **Soft-Impute** via fast Alternating Least Squares as described in [Matrix Completion and Low-Rank SVD via Fast Alternating Least Squares](https://arxiv.org/pdf/1410.2596.pdf).
    * <img src="https://raw.githubusercontent.com/rexyai/rsparse/master/docs/img/soft-impute.png" width="400">
    * with a solution in SVD form <img src="https://raw.githubusercontent.com/rexyai/rsparse/master/docs/img/soft-impute-svd-form.png" width="150">
1. **GloVe** as described in [GloVe: Global Vectors for Word Representation](https://nlp.stanford.edu/pubs/glove.pdf).
    * This is usually used to train word embeddings, but actually also very useful for recommender systems.
1. Matrix scaling as descibed in [EigenRec: Generalizing PureSVD for Effective and Efficient Top-N Recommendations](https://arxiv.org/pdf/1511.06033.pdf)

*********************

_Note: the optimized matrix operations which `rparse` used to offer have been moved to a [separate package](https://github.com/david-cortes/MatrixExtra)_

# Installation 

Most of the algorithms benefit from OpenMP and many of them could utilize high-performance implementations of BLAS. If you want to make the maximum out of this package, please read the section below carefully.

It is recommended to:

1. Use high-performance BLAS (such as OpenBLAS, MKL, Apple Accelerate).
1. Add proper compiler optimizations in your `~/.R/Makevars`. For example on recent processors (with AVX support) and compiler with OpenMP support, the following lines could be a good option:

```
CXX11FLAGS += -O3 -march=native -fopenmp
CXXFLAGS   += -O3 -march=native -fopenmp
```

### Mac OS

If you are on **Mac** follow the instructions at [https://mac.r-project.org/openmp/](https://mac.r-project.org/openmp/). After `clang` configuration, additionally put a `PKG_CXXFLAGS += -DARMA_USE_OPENMP` line in your `~/.R/Makevars`. After that, install `rsparse` in the usual way. 

Also we recommend to use [vecLib](https://developer.apple.com/documentation/accelerate/veclib) - Apple’s implementations of BLAS.

```sh
ln -sf  /System/Library/Frameworks/Accelerate.framework/Frameworks/vecLib.framework/Versions/Current/libBLAS.dylib /Library/Frameworks/R.framework/Resources/lib/libRblas.dylib
```

### Linux

On Linux, it's enough to just create this file if it doesn't exist (`~/.R/Makevars`).

If using OpenBLAS, it is highly recommended to use the `openmp` variant rather than the `pthreads` variant. On Linux, it is usually available as a separate package in typical distribution package managers (e.g. for Debian, it can be obtained by installing `libopenblas-openmp-dev`, which is not the default version), and if there are multiple BLASes installed, can be set as the default through the [Debian alternatives system](https://wiki.debian.org/DebianScience/LinearAlgebraLibraries) - which can also be used [for MKL](https://stackoverflow.com/a/49842944/5941695).

### Windows

By default, R for Windows comes with unoptimized BLAS and LAPACK libraries, and `rsparse` will prefer using Armadillo's replacements instead. In order to use BLAS, **install `rsparse` from source** (not from CRAN), removing the option `-DARMA_DONT_USE_BLAS` from `src/Makevars.win` and ideally adding `-march=native` (under `PKG_CXXFLAGS`). See [this tutorial](https://github.com/david-cortes/R-openblas-in-windows) for instructions on getting R for Windows to use OpenBLAS. Alternatively, Microsoft's MRAN distribution for Windows comes with MKL.

# Materials

**Note that syntax is these posts/slides is not up to date since package was under active development**

1. [Slides from DataFest Tbilisi(2017-11-16)](https://www.slideshare.net/DmitriySelivanov/matrix-factorizations-for-recommender-systems)
1. [Introduction to matrix factorization with Weighted-ALS algorithm](http://dsnotes.com/post/2017-05-28-matrix-factorization-for-recommender-systems/) - collaborative filtering for implicit feedback datasets.
1. [Music recommendations using LastFM-360K dataset](http://dsnotes.com/post/2017-06-28-matrix-factorization-for-recommender-systems-part-2/)
    * evaluation metrics for ranking
    * setting up proper cross-validation
    * possible issues with nested parallelism and thread contention
    * making recommendations for new users
    * complimentary item-to-item recommendations
1. [Benchmark](http://dsnotes.com/post/2017-07-10-bench-wrmf/) against other good implementations

Here is example of `rsparse::WRMF` on [lastfm360k](https://www.upf.edu/web/mtg/lastfm360k) dataset in comparison with other good implementations:

<img src="https://github.com/dselivanov/bench-wals/raw/master/img/wals-bench-cg.png" width="600">


# API

We follow [mlapi](https://github.com/dselivanov/mlapi) conventions.

# Release and configure

## Making release

Don't forget to add `DARMA_NO_DEBUG` to `PKG_CXXFLAGS` to skip bound checks (this has significant impact on NNLS solver)

```
PKG_CXXFLAGS = ... -DARMA_NO_DEBUG
```

## Configure

Generate configure:

```sh
autoconf configure.ac > configure && chmod +x configure
```


NEWS.md
# rsparse dev
- faster WRMF solver see #72, #75
- updated github actions

# rsparse 0.5.1 (2022-09-11)
- update `configure` script, thanks to @david-cortes, see #73
- minor fixes in WRMF
- update docs with new roxygen2 to pass CRAN checks
- update NEWS.md ro follow CRAN format

# rsparse 0.5.0 (2021-10-17)
- reworked non-negative matrix factorization with brand-new Coordinate Descent solver for OLS
- WRMF can model user, item and global biases
- various performance improvements

# rsparse 0.4.0 (2020-04-01)
- updated docs with roxygen2 7.1
- added `ScaleNormalize` transformer
- added sparse*float S4 methods

# rsparse 0.3.3.2 (2019-07-17)
- faster `find_top_product()` - avoid BLAS and openmp thread contention
- correctly identify openmp on OSX
- fixed issue with CRAN 'rcnst' check
- use `install_name_tool` hook in the `.onLoad()` - changes location of the `float.so` for CRAN binary installation - see #25

# rsparse 0.3.3.1 (2019-04-14)
- fixed out of bound memory access as reported by CRAN UBSAN
- added ability to init GloVe embeddings with user provided values

# rsparse 0.3.3 (2019-03-16)
- added methods to natively slice CSR matrices without converting them to triplet/CSC
- add GloVe matrix factorization (adapted from `text2vec`)
- link to `float` package - credits to @snoweye and @wrathematics


R/ALS.R
#' @name ALS
#'
#' @title ALS implicit
#' @description Creates matrix factorization model which could be solved with weighted or vanilla Alternating Least Squares algorithm.
#' For implicit feedback see (Hu, Koren, Volinsky)'2008 paper \url{http://yifanhu.net/PUB/cf.pdf}.
#' For explicit feedback model is classic model for rating matrix decomposition with MSE error (without biases at the moment).
#' @seealso
#' \itemize{
#'   \item{\url{https://math.stackexchange.com/questions/1072451/analytic-solution-for-matrix-factorization-using-alternating-least-squares/1073170#1073170}}
#'   \item{\url{http://activisiongamescience.github.io/2016/01/11/Implicit-Recommender-Systems-Biased-Matrix-Factorization/}}
#'   \item{\url{http://datamusing.info/blog/2015/01/07/implicit-feedback-and-collaborative-filtering/}}
#'   \item{\url{https://jessesw.com/Rec-System/}}
#'   \item{\url{http://danielnee.com/2016/09/collaborative-filtering-using-alternating-least-squares/}}
#'   \item{\url{http://www.benfrederickson.com/matrix-factorization/}}
#'   \item{\url{http://www.benfrederickson.com/fast-implicit-matrix-factorization/}}
#' }
#' @format \code{R6Class} object.
#' @section Usage:
#' For usage details see \bold{Methods, Arguments and Examples} sections.
#' \preformatted{
#'   model = ALS$new(rank = 10L, lambda = 0,
#'                   feedback = c("implicit", "explicit"),
#'                   init_stdv = 0.01,
#'                   n_threads = parallel::detectCores(),
#'                   non_negative = FALSE,
#'                   solver = c("conjugate_gradient", "cholesky"),
#'                   cg_steps = 3L)
#'   model$fit_transform(x, n_iter = 5L, n_threads = 1, ...)
#'   model$predict(x, k, n_threads = private$n_threads, not_recommend = x, ...)
#'   model$components
#'   model$add_scorer(x, y, name, metric, ...)
#'   model$remove_scorer(name)
#'   model$ap_k(x, y, k = ncol(x))
#'   model$ndcg_k(x, y, k = ncol(x))
#' }
#' @section Methods:
#' \describe{
#'   \item{\code{$new(rank = 10L, lambda = 0, feedback = c("implicit", "explicit"),
#'                    init_stdv = 0.01, n_threads = parallel::detectCores(), non_negative = FALSE,
#'                    solver = c("conjugate_gradient", "cholesky"), cg_steps = 3L) }}{ creates matrix
#'     factorization model model with \code{rank} latent factors}
#'   \item{\code{$fit_transform(x, n_iter = 5L, n_threads = private$n_threads, ...)}}{ fits model to
#'     an input user-item matrix. (preferably in "dgCMatrix" format).
#'     For implicit feedback \code{x} should be a confidence matrix which corresponds to \code{1 + alpha * r_ui} in original paper.
#'     Usually \code{r_ui} corresponds to the number of interactions of user \code{u} and item \code{i}.
#'     For explicit feedback values in \code{x} represents ratings.
#'     \bold{Returns factor matrix for users of size \code{n_users * rank}}}
#'   \item{\code{$predict(x, k, n_threads = private$n_threads, not_recommend = x, ...)}}{predict \code{top k}
#'     item ids for users \code{x} (= column names from the matrix passed to \code{fit_transform()} method).
#'     Users features should be defined the same way as they were defined in training data - as \bold{sparse matrix}
#'     of confidence values (implicit feedback) or ratings (explicit feedback).
#'     Column names (=item ids) should be in the same order as in the \code{fit_transform()}.}
#'   \item{\code{$add_scorer(x, y, name, metric, ...)}}{add a metric to watchlist.
#'   Metric will be evaluated after each ALS interation. At the moment following metrices are supported:
#'     \bold{"loss"}, \bold{"map@@k"}, \bold{"ndcg@@k"}, where \bold{k} is some integer. For example \code{map@@10}.}
#'   \item{\code{$remove_scorer(name)}}{remove a metric from watchlist}
#'   \item{\code{$ap_k(x, y, k = ncol(x)}}{ calculates average precision at k for new/out-of-bag users \code{x} and
#'   validate against ground truth values in \code{y}}
#'   \item{\code{$ndcg_k(x, y, k = ncol(x)}}{ calculates NDCG k for new/out-of-bag users \code{x} and
#'   validate against ground truth values in \code{y}}
#'   \item{\code{$components}}{item factors matrix of size \code{rank * n_items}}
#'}
#' @section Arguments:
#' \describe{
#'  \item{model}{A \code{ALS} model.}
#'  \item{x}{An input sparse user-item matrix(of class \code{dgCMatrix}).
#'  For explicit feedback should consists of ratings.
#'  For implicit feedback all positive interactions should be filled with \bold{confidence} values.
#'  Missed interactions should me zeros/empty.
#'  So for simple case case when \code{confidence = 1 + alpha * x}}
#'  \item{y}{An input user-item \bold{relevance} matrix. Used during evaluation of \code{map@@k}, \code{ndcg@@k}
#'    Should have the same shape as corresponding confidence matrix \code{x}.
#'    Values are used as "relevance" in ndgc calculation}
#'  \item{name}{\code{character} - user-defined name of the scorer. For example "ndcg-scorer-1"}
#'  \item{rank}{\code{integer} - number of latent factors}
#'  \item{lambda}{\code{numeric} - regularization parameter}
#'  \item{feedback}{\code{character} - feedback type - one of \code{c("implicit", "explicit")}}
#'  \item{solver}{\code{character} - solver for "implicit feedback" problem.
#'     One of \code{c("conjugate_gradient", "cholesky")}.
#'     Usually approximate \code{"conjugate_gradient"} is significantly faster and solution is
#'     on par with exact \code{"cholesky"}}
#'  \item{cg_steps}{\code{integer > 0} - max number of internal steps in conjugate gradient
#'     (if "conjugate_gradient" solver used). \code{cg_steps = 3} by default.
#'     Controls precision of linear equation solution at the each ALS step. Usually no need to tune this parameter.}
#'  \item{n_threads}{\code{numeric} default number of threads to use during training and prediction
#'  (if OpenMP is available). Check descriptions of other methods - some of them can override this parameter.}
#'  \item{not_recommend}{\code{sparse matrix} or \code{NULL} - points which items should be excluided from recommendations for a user.
#'    By default it excludes previously seen/consumed items.}
#'  \item{convergence_tol}{{\code{numeric = -Inf} defines early stopping strategy. We stop fitting
#'     when one of two following conditions will be satisfied: (a) we have used
#'     all iterations, or (b) \code{loss_previous_iter / loss_current_iter - 1 < convergence_tol}}}
#'  \item{init_stdv}{\code{numeric} standart deviation for initialization of the initial latent matrices}
#'  \item{...}{other arguments. Not used at the moment}
#' }
#' @export
ALS = R6::R6Class(
  inherit = mlapi::mlapiDecomposition,
  classname = "AlternatingLeastSquares",
  public = list(
    initialize = function(rank = 10L,
                          lambda = 0,
                          feedback = c("implicit", "explicit"),
                          n_threads = parallel::detectCores(),
                          init_stdv = 0.01,
                          non_negative = FALSE,
                          solver = c("conjugate_gradient", "cholesky"),
                          cg_steps = 3L) {

      solver = match.arg(solver)
      private$feedback = match.arg(feedback)
      if(solver == "conjugate_gradient" && private$feedback == "explicit")
        flog.warn("only 'cholesky' is available for 'explicit' feedback")

      if(solver == "cholesky") private$solver_code = 0L
      if(solver == "conjugate_gradient") private$solver_code = 1L

      if(solver == "cholesky") flog.info("'cg_steps' ignored for 'cholesky' solver")
      stopifnot(is.integer(cg_steps) && length(cg_steps) == 1)
      private$cg_steps = cg_steps

      private$set_internal_matrix_formats(sparse = "dgCMatrix", dense = NULL)
      private$lambda = lambda
      private$init_stdv = init_stdv
      private$rank = rank

      private$scorers = new.env(hash = TRUE, parent = emptyenv())
      private$n_threads = n_threads
      private$non_negative = non_negative
    },
    fit_transform = function(x, n_iter = 5L, convergence_tol = -Inf, n_threads = private$n_threads, ...) {

      # x = confidense matrix, not ratings/interactions matrix!
      # we expect user already transformed it
      # default choice will be
      # x = 1 + alpha * r

      flog.debug("convert input to %s if needed", private$internal_matrix_formats$sparse)
      c_ui = private$check_convert_input(x, private$internal_matrix_formats)

      # strore item_ids in order to use them in predict method
      private$item_ids = colnames(c_ui)

      flog.debug("check items in input are not negative")
      stopifnot(all(c_ui@x >= 0))

      flog.debug("making antoher matrix for convenient traverse by users - transposing input matrix")
      c_iu = t(c_ui)

      # init
      n_user = nrow(c_ui)
      n_item = ncol(c_ui)

      # private$U = matrix(rnorm(n_user * private$rank, 0, private$init_stdv), ncol = n_user, nrow = private$rank)
      private$U = matrix(0.0, ncol = n_user, nrow = private$rank)
      private$I = matrix(rnorm(n_item * private$rank, 0, private$init_stdv), ncol = n_item, nrow = private$rank)
      Lambda = diag(x = private$lambda, nrow = private$rank, ncol = private$rank)

      trace_values = vector("numeric", n_iter)

      flog.info("starting factorization with %d threads", n_threads)
      trace_lst = vector("list", n_iter)
      loss_prev_iter = Inf
      # iterate
      for (i in seq_len(n_iter)) {

        flog.debug("iter %d by item", i)
        stopifnot(ncol(private$U) == ncol(c_iu))
        if (private$feedback == "implicit") {
          # private$U will be modified in place
          loss = als_implicit(c_iu, private$I, private$U, n_threads = n_threads,
                       # pass -1 mean that we ask to not calculate loss
                       # 0 loss will be returned instead
                       lambda = -1,
                       solver = private$solver_code, cg_steps = private$cg_steps)
          # private$U = private$solver(private$I, c_iu, n_threads = n_threads, ...)
        } else if (private$feedback == "explicit") {
          private$U = private$solver_explicit_feedback(c_iu, private$I)
        }
        # if need non-negative matrix factorization - just set all negative values to zero
        if(private$non_negative)
          private$U[private$U < 0] = 0

        flog.debug("iter %d by user", i)
        stopifnot(ncol(private$I) == ncol(c_ui))
        if (private$feedback == "implicit") {
          # private$I will be modified in place
          loss = als_implicit(c_ui, private$U, private$I, n_threads = n_threads,
                              lambda = private$lambda,
                              private$solver_code, private$cg_steps)
          # private$I = private$solver(private$U, c_ui, n_threads = n_threads, ...)
        } else if (private$feedback == "explicit") {
          private$I = private$solver_explicit_feedback(c_ui, private$U)
        }
        # if need non-negative matrix factorization - just set all negative values to zero
        if(private$non_negative)
          private$I[private$I < 0] = 0

        #------------------------------------------------------------------------
        # calculate some metrics if needed in order to diagnose convergence
        #------------------------------------------------------------------------
        if (private$feedback == "explicit")
          loss = als_loss_explicit(c_ui, private$U, private$I, private$lambda, n_threads);

        trace_iter = vector("list", length(names(private$scorers)))
        j = 1L
        trace_scors_string = ""
        for(sc in names(private$scorers)) {
          score = private$scorers[[sc]]()
          trace_scors_string = sprintf("%s score %s = %f", trace_scors_string, sc, score)
          trace_iter[[j]] = list(iter = i, scorer = sc, value = score)
          j = j + 1L
        }
        trace_iter = data.table::rbindlist(trace_iter)
        trace_lst[[i]] = data.table::rbindlist(list(trace_iter, list(iter = i, scorer = "loss", value = loss)))
        flog.info("iter %d loss = %.4f %s", i, loss, trace_scors_string)
        if(loss_prev_iter / loss - 1 < convergence_tol) {
          flog.info("Converged after %d iterations", i)
          break
        }
        loss_prev_iter = loss
        #------------------------------------------------------------------------
      }

      private$components_ = private$I
      data.table::setattr(private$components_, "dimnames", list(NULL, colnames(x)))

      res = t(private$U)
      setattr(res, "trace", rbindlist(trace_lst))
      setattr(res, "dimnames", list(rownames(x), NULL))
      res
    },
    # project new users into latent user space - just make ALS step given fixed items matrix
    transform = function(x, n_threads = private$n_threads, ...) {
      stopifnot(ncol(x) == ncol(private$I))
      # allocate result matrix - will be modified in place

      x = private$check_convert_input(x, private$internal_matrix_formats)

      if(private$feedback == "implicit") {
        res = matrix(0, nrow = private$rank, ncol = nrow(x))

        als_implicit(t(x), private$I, res, n_threads = n_threads,
                     lambda = private$lambda,
                     private$solver_code, private$cg_steps)

      } else if(private$feedback == "explicit")
        res = private$solver_explicit_feedback(t(x), private$I)
      else
        stop(sprintf("don't know how to work with feedback = '%s'", private$feedback))
      if(private$non_negative)
        res[res < 0] = 0
      res = t(res)
      data.table::setattr(res, "dimnames", list(rownames(x), NULL))
      res
    },
    # project new items into latent item space
    get_items_embeddings = function(x, n_threads = private$n_threads, ...) {
      if(private$feedback == "implicit") {
        res = matrix(0, nrow = private$rank, ncol = nrow(x))
        als_implicit(x, private$U, res, n_threads = n_threads,
                     lambda = private$lambda,
                     private$solver_code, private$cg_steps)
      } else if(private$feedback == "explicit") {
        res = private$solver_explicit_feedback(x, private$U)
      } else
        stop(sprintf("don't know how to work with feedback = '%s'", private$feedback))

      if(private$non_negative)
        res[res < 0] = 0

      res
    },
    predict = function(x, k, n_threads = private$n_threads, not_recommend = x, ...) {
      stopifnot(private$item_ids == colnames(x))
      stopifnot(is.null(not_recommend) || inherits(not_recommend, "sparseMatrix"))
      if(!is.null(not_recommend))
        not_recommend = as(not_recommend, "dgCMatrix")
      m = nrow(x)

      # transform user features into latent space
      # calculate scores for each item
      user_item_score = self$transform(x) %*% private$I
      indices = top_k_indices_byrow(user_item_score, not_recommend, k, n_threads)
      scores = attr(indices, "scores", exact = TRUE)
      attr(indices, "scores") = NULL

      predicted_item_ids = private$item_ids[indices]
      data.table::setattr(predicted_item_ids, "dim", dim(indices))
      data.table::setattr(predicted_item_ids, "indices", indices)
      data.table::setattr(predicted_item_ids, "scores", scores)
      data.table::setattr(predicted_item_ids, "dimnames", list(rownames(x), NULL))
      predicted_item_ids
    },
    ap_k = function(x, y, k = ncol(x), ...) {
      stopifnot(ncol(x) == ncol(y))
      stopifnot(nrow(x) == nrow(y))
      n_u = nrow(x)
      preds = self$predict(x, k, ...)
      preds = attr(preds, "indices", TRUE)
      y_csr = as(y, "RsparseMatrix")
      res = numeric(n_u)
      for(u in seq_len(n_u)) {
        p1 = y_csr@p[[u]]
        p2 = y_csr@p[[u + 1]]
        ind = p1 + seq_len(p2 - p1)
        u_ind = y_csr@j[ind] + 1L
        u_x = y_csr@x[ind]
        ord = order(u_x, decreasing = TRUE)
        res[[u]] = ap_at_k(preds[u, ], u_ind[ord], k = k)
      }
      res
    },
    ndcg_k = function(x, y, k = ncol(x), ...) {
      # stopifnot(ncol(x) == ncol(y))
      stopifnot(nrow(x) == nrow(y))
      n_u = nrow(x)
      preds = self$predict(x, k, ...)
      preds = attr(preds, "indices", TRUE)
      y_csr = as(y, "RsparseMatrix")
      res = numeric(n_u)
      for(u in seq_len(n_u)) {
        p1 = y_csr@p[[u]]
        p2 = y_csr@p[[u + 1]]
        ind = p1 + seq_len(p2 - p1)
        u_ind = y_csr@j[ind] + 1L
        u_x = y_csr@x[ind]
        ord = order(u_x, decreasing = TRUE)
        res[[u]] = ndcg_at_k(preds[u, ], u_ind[ord], u_x[ord], k)
      }
      res
    },
    # x = sparse matrix of observed user interactions
    # y = sparse matrix of observed user we are trying to predtict
    add_scorer = function(x, y, name, metric, ...) {
      if(exists(name, where = private$scorers, inherits = FALSE))
        stop(sprintf("scorer with name '%s' already exists", name))

      if(metric == "loss") {
        private$scorers[[name]] = function() {
          # transpose since internally users kept as n_factor * n_users
          U_new = t(self$transform(x))
          calc_als_implicit_loss(x, private$U, private$I, private$lambda, n_threads = private$n_threads, ...)
        }
      } else {
        if (length(grep(pattern = "(ndcg|map)\\@[[:digit:]]+", x = metric)) != 1 )
          stop(sprintf("don't know how add '%s' metric. Only 'loss', 'map@k', 'ndcg@k' are supported", metric))

        scorer_conf = strsplit(metric, "@", T)[[1]]
        k = as.integer(tail(scorer_conf, 1))
        scorer_fun = scorer_conf[[1]]

        if(scorer_fun == "map")
          private$scorers[[name]] = function() mean(self$ap_k(x, y, k, ...), na.rm = T)
        if(scorer_fun == "ndcg")
          private$scorers[[name]] = function() mean(self$ndcg_k(x, y, k, ...), na.rm = T)
      }
    },
    remove_scorer = function(name) {
      if(!exists(name, where = private$scorers))
        stop(sprintf("can't find scorer '%s'", name))
      rm(list = name, envir = private$scorers)
    },
    finalize = function() {
      rm(list = names(private$scorers), envir = private$scorers)
      private$scorers = NULL
      gc();
    }
  ),
  private = list(
    solver_code = NULL,
    cg_steps = NULL,
    scorers = NULL,
    lambda = NULL,
    init_stdv = NULL,
    rank = NULL,
    n_threads = NULL,
    non_negative = NULL,
    # user factor matrix = rank * n_users
    U = NULL,
    # item factor matrix = rank * n_items
    I = NULL,
    feedback = NULL,
    item_ids = NULL,
    #------------------------------------------------------------
    solver_explicit_feedback = function(R, X) {
      XtX = tcrossprod(X) + diag(x = private$lambda, nrow = private$rank, ncol = private$rank)
      solve(XtX, as(X %*% R, "matrix"))
    },
    # X = factor matrix n_factors * (n_users or n_items)
    # C_UI = user-item confidence matrix

    # R solver for reference. Now replaced with fast Armadillo solver
    solver = function(X, C_UI, n_threads = private$n_threads, ...) {

      XtX_reg = tcrossprod(X) + diag(x = private$lambda, nrow = private$rank, ncol = private$rank)
      m_rank = nrow(X)

      # BLAS multithreading should be switched off
      # https://stat.ethz.ch/pipermail/r-sig-hpc/2012-July/001432.html
      # https://hyperspec.wordpress.com/2012/07/26/altering-openblas-threads/
      # https://stat.ethz.ch/pipermail/r-sig-debian/2016-August/002586.html

      nc = ncol(C_UI)

      # aliases to sparse matrix components for convenience
      C_UI_P = C_UI@p
      C_UI_I = C_UI@i
      C_UI_X = C_UI@x

      splits = parallel::splitIndices(nc, n_threads)

      RES = parallel::mclapply(splits, function(chunk) {
        # MAP PHASE
        chunk_len = length(chunk)
        chunk_start = chunk[[1]]
        chunk_end = chunk[[chunk_len]]
        n_print = as.integer(chunk_len / 16)

        RES_WORKER = matrix(data = 0, nrow = m_rank, ncol = chunk_len)

        # RES_WORKER = lapply(chunk, function(i) {

        for(j in seq_along(chunk)) {
          i = chunk[[j]]
          if(isTRUE(i %% n_print == 0)) {
            flog.debug("worker %d progress %d%%", Sys.getpid(), as.integer(100 * (i - chunk_start) / chunk_len))
          }

          # column pointers
          p1 = C_UI_P[[i]]
          p2 = C_UI_P[[i + 1L]]
          pind = p1 + seq_len(p2 - p1)

          # add 1L because in C_UI indices are 0-based and in dense matrices they are 1-based
          ind = C_UI_I[pind] + 1L
          confidence = C_UI_X[pind]

          # corresponds to Y for a given user by we take only columns which have non-zero interactions
          # drop = FALSE - don't simplify matrices with 1 column to a vector
          X_nnz = X[, ind, drop = FALSE]

          # line below corresponds to [(YtY + lambda * I) + Yt %*% (C_u - I) %*% Y] in paper Hu, Koren, Volinsky paper
          # XtX_reg = (YtY + lambda * I)
          # confidence - 1 = C_u - I
          # X_nnz %*% (t.default(X_nnz) * (confidence - 1)) = Yt %*% (C_u - I) %*% Y

          inv = XtX_reg + X_nnz %*% (t.default(X_nnz) * (confidence - 1))
          rhs = X_nnz %*% confidence


          # same as RES_WORKER[, j] = solve(inv, rhs)
          # but doesn't perform many checks, doesn't copy attributes, etc. So it is notably faster.
          RES_WORKER[, j] = .Internal(La_solve(inv, rhs, .Machine$double.eps))

          #------------------------------------
          # what didn't work
          #------------------------------------
          # This was slower
          # X_nnz = t.default(X[, ind, drop = FALSE])
          # inv = XtX_reg + base::crossprod(X_nnz, X_nnz * (confidence - 1))
          # rhs = base::crossprod(X_nnz, confidence)

          # Cholesky was slower as well
          # RES_WORKER[, j] = chol2inv(chol.default(inv)) %*% rhs
          #------------------------------------
        }
        RES_WORKER
      }, mc.cores = n_threads, ...)
      # GLOBAL REDUCE
      do.call(cbind, RES)
    }
  )
)
